File: go/core/id/contract_test.go
package id_test

// Cross-language parity contract test for the typeid primitive.
//
// Loads contracts/typeid-v1/fixtures.json (single source of truth shared
// with the Rust, TypeScript, Python and PHP SDKs) and asserts that this
// Go SDK produces and parses the canonical typeid strings byte-for-byte.
//
// A failure here means either:
//   - the upstream go.jetify.com/typeid encoder drifted, or
//   - the contract fixtures themselves were edited without updating the
//     other-language SDKs.
// Either way the parity matrix is broken; fix the encoder or coordinate
// a fixtures.json + 5-SDK bump (tlc T-0753).

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"slices"
	"testing"

	"github.com/google/uuid"
	typeid "go.jetify.com/typeid"

	"hop.top/kit/go/core/id"
)

type contractVector struct {
	Name   string   `json:"name"`
	Prefix string   `json:"prefix"`
	UUID   string   `json:"uuid"`
	TypeID string   `json:"typeid"`
	SkipIn []string `json:"skip_in,omitempty"`
	Note   string   `json:"note,omitempty"`
}

type contractFile struct {
	Version string           `json:"version"`
	Spec    string           `json:"spec"`
	Vectors []contractVector `json:"vectors"`
}

// loadContract finds the repo root by walking up from this test file's
// own location until contracts/typeid-v1/fixtures.json appears. Avoids
// embed-style directives (which forbid paths outside the package
// directory) and CWD-relative reads (fragile under `go test ./...`).
func loadContract(t *testing.T) contractFile {
	t.Helper()
	_, here, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller(0) failed; cannot locate contracts/typeid-v1/fixtures.json")
	}
	dir := filepath.Dir(here)
	for i := 0; i < 10; i++ {
		candidate := filepath.Join(dir, "contracts", "typeid-v1", "fixtures.json")
		if _, err := os.Stat(candidate); err == nil {
			raw, err := os.ReadFile(candidate)
			if err != nil {
				t.Fatalf("read %s: %v", candidate, err)
			}
			var cf contractFile
			if err := json.Unmarshal(raw, &cf); err != nil {
				t.Fatalf("parse %s: %v", candidate, err)
			}
			return cf
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
		dir = parent
	}
	t.Fatal("contracts/typeid-v1/fixtures.json: not found walking up from test file")
	return contractFile{}
}

func TestContract_Version(t *testing.T) {
	cf := loadContract(t)
	if cf.Version != "v1" {
		t.Errorf("contract version: want v1, got %q", cf.Version)
	}
	if cf.Spec != "jetify-typeid-v0.3" {
		t.Errorf("contract spec: want jetify-typeid-v0.3, got %q", cf.Spec)
	}
	if len(cf.Vectors) == 0 {
		t.Fatal("contract has no vectors")
	}
}

// TestContract_Generation walks every vector and verifies that the
// upstream encoder (the same one the kit's New() path delegates to)
// produces the pinned canonical typeid string for the pinned UUID. This
// is the "wire format hasn't drifted" guard for the Go SDK.
func TestContract_Generation(t *testing.T) {
	cf := loadContract(t)
	for _, v := range cf.Vectors {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			if slices.Contains(v.SkipIn, "go") {
				t.Skipf("skip_in includes 'go': %s", v.Note)
			}
			tid, err := typeid.FromUUIDWithPrefix(v.Prefix, v.UUID)
			if err != nil {
				t.Fatalf("typeid.FromUUIDWithPrefix(%q, %q): %v", v.Prefix, v.UUID, err)
			}
			if got := tid.String(); got != v.TypeID {
				t.Errorf("canonical typeid drift:\n got:  %q\n want: %q\n(prefix=%q uuid=%q)",
					got, v.TypeID, v.Prefix, v.UUID)
			}
		})
	}
}

// TestContract_Parse exercises the kit's public Parse() against every
// vector. The (prefix, uuid) recovered from the canonical string must
// equal the pinned input.
func TestContract_Parse(t *testing.T) {
	cf := loadContract(t)
	for _, v := range cf.Vectors {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			if slices.Contains(v.SkipIn, "go") {
				t.Skipf("skip_in includes 'go': %s", v.Note)
			}
			parsed, err := id.Parse(v.TypeID)
			if err != nil {
				t.Fatalf("id.Parse(%q): %v", v.TypeID, err)
			}
			if parsed.Prefix != v.Prefix {
				t.Errorf("prefix mismatch: want %q, got %q", v.Prefix, parsed.Prefix)
			}
			wantUUID, err := uuid.Parse(v.UUID)
			if err != nil {
				t.Fatalf("uuid.Parse fixture %q: %v", v.UUID, err)
			}
			if parsed.UUID != wantUUID {
				t.Errorf("uuid mismatch: want %s, got %s", wantUUID, parsed.UUID)
			}
		})
	}
}

File: sdk/ts/src/id/contract.test.ts
// Cross-language parity contract test (tlc T-0753).
//
// Loads contracts/typeid-v1/fixtures.json from the repo root and
// asserts that this TS SDK's encode (via typeid-js's TypeID.fromUUID)
// and decode (via the kit's parse) agree with the canonical wire form
// shared by go/rs/py/php. A divergence here means either typeid-js
// drifted or the contract was edited without updating all five SDKs.

import { existsSync, readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, it, expect } from 'vitest';
import { TypeID } from 'typeid-js';

import { parse } from './index';

interface ContractVector {
  readonly name: string;
  readonly prefix: string;
  readonly uuid: string;
  readonly typeid: string;
  readonly skip_in?: ReadonlyArray<string>;
  readonly note?: string;
}

interface ContractFile {
  readonly version: string;
  readonly spec: string;
  readonly vectors: ReadonlyArray<ContractVector>;
}

// Walk up from this test file's directory until we hit a directory
// holding contracts/typeid-v1/fixtures.json. Robust against both
// `vitest run` (CWD = sdk/ts) and IDE-driven invocations whose CWD
// may vary.
function locateContract(): string {
  const here = dirname(fileURLToPath(import.meta.url));
  let dir = here;
  for (let i = 0; i < 10; i++) {
    const candidate = resolve(dir, 'contracts', 'typeid-v1', 'fixtures.json');
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  throw new Error(
    `contracts/typeid-v1/fixtures.json: not found walking up from ${here}`,
  );
}

const CONTRACT: ContractFile = JSON.parse(
  readFileSync(locateContract(), 'utf8'),
) as ContractFile;

describe('typeid-v1 contract metadata', () => {
  it('pins version + spec', () => {
    expect(CONTRACT.version).toBe('v1');
    expect(CONTRACT.spec).toBe('jetify-typeid-v0.3');
    expect(CONTRACT.vectors.length).toBeGreaterThan(0);
  });
});

describe('typeid-v1 contract generation (TS)', () => {
  for (const v of CONTRACT.vectors) {
    const skipped = v.skip_in?.includes('ts') ?? false;
    const fn = skipped ? it.skip : it;
    fn(`vector ${v.name}: TypeID.fromUUID produces the pinned canonical`, () => {
      const got = TypeID.fromUUID(v.prefix, v.uuid).toString();
      expect(got).toBe(v.typeid);
    });
  }
});

describe('typeid-v1 contract parse (TS)', () => {
  for (const v of CONTRACT.vectors) {
    const skipped = v.skip_in?.includes('ts') ?? false;
    const fn = skipped ? it.skip : it;
    fn(`vector ${v.name}: parse round-trips to (prefix, uuid)`, () => {
      const parsed = parse(v.typeid);
      expect(parsed.prefix).toBe(v.prefix);
      expect(parsed.uuid).toBe(v.uuid);
    });
  }
});
